source/solvers/postprocessing_velocities.cc

```diff
@@ -43,18 +54,121 @@ AverageVelocities<dim, VectorType, DofsType>::calculate_average_velocities(
   // total time = 0.
   inv_range_time = 1. / ((current_time - real_initial_time) + dt_0);
 
-  // Calculate average_velocities. (u*dt) / (total time + dt)
-  // The sum of all weighted velocities in time / total time calculated
-  average_velocities.equ(inv_range_time, sum_velocity_dt);
+
+  // If calculate reynolds stress is enable, average velocities must be
+  // calculated at every iteration since the average values are needed to get
+  // reynolds stress. If it is disabled, average velocities only need to be
+  // calculated when output is required.
+  if (post_processing.calculate_reynolds_stress)
+    {
+      average_velocities.equ(inv_range_time, sum_velocity_dt);
+
+      this->calculate_reynolds_stress(local_evaluation_point,
+                                      is_output_iteration);
+    }
+  else if (is_output_iteration)
+    {
+      // Calculate average_velocities. (u*dt) / (total time + dt)
+      // The sum of all weighted velocities in time / total time calculated
+      average_velocities.equ(inv_range_time, sum_velocity_dt);
+    }
 }
 
+// Since Trilinos vectors and block vectors data doesn't have the same
+// structure, those vectors need to be processed in different ways.
+// Same about the dimension, where 2D solutions are stored in (u,v,p) groups
+// for vectors and [(u,v)][p] for block vectors and where 3D solutions
+// are stored in (u,v,w,p) groups for vectors and [(u,v,w)][p] for
+// block vectors.
 template <int dim, typename VectorType, typename DofsType>
-const VectorType
-AverageVelocities<dim, VectorType, DofsType>::get_average_velocities()
+void
+AverageVelocities<dim, VectorType, DofsType>::calculate_reynolds_stress(
+  const VectorType &local_evaluation_point,
+  const bool &      is_output_iteration)
 {
-  return average_velocities;
+  if constexpr (std::is_same_v<VectorType, TrilinosWrappers::MPI::Vector>)
+    {
+      const unsigned int begin_index =
+        local_evaluation_point.local_range().first;
+      const unsigned int end_index =
+        local_evaluation_point.local_range().second;
+
+      for (unsigned int i = begin_index; i < end_index; i++)
```

Comment:
Why loop with increments of one then? You could loop by increments of dim+1 : " i+= dim+1". C++ allows thats. You can do whatever you want to i.

Reply:
Because if begin_index is not 0, dim+1, 2(dim+1), ... i will never get the right value for the vector indexation. I know it's kinda weird, it's because of the <u'v'> calculation. I tried it and it didn't work, but I will try again to make sure.

Reply:
I did try and it did not work, but I must have done something wrong because it actually works mhm.